src/type/validate.ts

```diff
@@ -196,8 +214,126 @@ function validateDirectives(context: SchemaValidationContext): void {
+  inputValue: GraphQLArgument | GraphQLInputField,
+): void {
+  const defaultValue = inputValue.defaultValue;
+
+  if (!defaultValue) {
+    return;
+  }
+
+  if (defaultValue.literal) {
+    validateInputLiteral(
+      defaultValue.literal,
+      inputValue.type,
+      undefined,
+      (error, path) => {
+        context.reportError(
+          `${inputValue} has invalid default value${printPathArray(path)}: ${
+            error.message
+          }`,
+          error.nodes,
+        );
+      },
+    );
+  } else {
+    const errors: Array<[GraphQLError, ReadonlyArray<string | number>]> = [];
+    validateInputValue(defaultValue.value, inputValue.type, (error, path) => {
+      errors.push([error, path]);
+    });
+
+    // If there were validation errors, check to see if it can be "uncoerced"
+    // and then correctly validated. If so, report a clear error with a path
+    // to resolution.
+    if (errors.length > 0) {
+      try {
+        const uncoercedValue = uncoerceDefaultValue(
+          defaultValue.value,
+          inputValue.type,
+        );
+
+        const uncoercedErrors = [];
+        validateInputValue(uncoercedValue, inputValue.type, (error, path) => {
+          uncoercedErrors.push([error, path]);
+        });
+
+        if (uncoercedErrors.length === 0) {
+          context.reportError(
+            `${inputValue} has invalid default value: ${inspect(
+              defaultValue.value,
+            )}. Did you mean: ${inspect(uncoercedValue)}?`,
+            inputValue.astNode?.defaultValue,
+          );
+          return;
+        }
+      } catch (_error) {
+        // ignore
+      }
+    }
+
+    // Otherwise report the original set of errors.
+    for (const [error, path] of errors) {
+      context.reportError(
+        `${inputValue} has invalid default value${printPathArray(path)}: ${
+          error.message
+        }`,
+        inputValue.astNode?.defaultValue,
+      );
+    }
+  }
+}
+
+/**
+ * Historically GraphQL.js allowed default values to be provided as
+ * assumed-coerced "internal" values, however default values should be provided
+ * as "external" pre-coerced values. `uncoerceDefaultValue()` will convert such
+ * "internal" values to "external" values to display as part of validation.
+ *
+ * This performs the "opposite" of `coerceInputValue()`. Given an "internal"
+ * coerced value, reverse the process to provide an "external" uncoerced value.
+ */
+function uncoerceDefaultValue(value: unknown, type: GraphQLInputType): unknown {
+  if (isNonNullType(type)) {
+    return uncoerceDefaultValue(value, type.ofType);
+  }
+
+  if (value === null) {
+    return null;
+  }
+
+  if (isListType(type)) {
+    if (isIterableObject(value)) {
+      return Array.from(value, (itemValue) =>
+        uncoerceDefaultValue(itemValue, type.ofType),
+      );
     }
+    return [uncoerceDefaultValue(value, type.ofType)];
   }
+
+  if (isInputObjectType(type)) {
+    invariant(isObjectLike(value));
```

Comment:
Should probably have a message?

Reply:
I think @leebyron does not include a message because any errors thrown by this function are caught and discarded. If (a) uncoercing succeeds and (b) validation of the uncoerced value results in no errors, we will raise an error suggesting that an internal value was incorrectly supplied (as per the old behavior) rather than an external value.